`alloc`: add unstable cfg features `no_rc` and `no_sync`

ojeda opened the pull request:
In Rust for Linux we are using these to make `alloc` a bit more modular.

See https://github.com/rust-lang/rust/pull/86048 and https://github.com/rust-lang/rust/pull/84266 for similar requests.

Of course, the particular names are not important.

ojeda committed:
`alloc`: add unstable cfg features `no_rc` and `no_sync`

In Rust for Linux we are using these to make `alloc` a bit
more modular.

A `run-make-fulldeps` test is added for each of them, so that
enabling each of them independently is kept in a compilable state.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

kennytm commented:
> Do we want to keep all possible combinations working and maintained if we keep adding this sort of flags (in the extreme, for every module)?

The `no_rc` and `no_sync` cfg features as-is which any tests is like a Tier-3 promise. Reviewers can't possibly figure out if all these combinations are working, so those responsible for these features have to fix them up after the fact.

The [`no_global_oom_handling`](https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/alloc-no-oom-handling) (#84266) and [`no_fp_fmt_parse`](https://github.com/rust-lang/rust/blob/master/src/test/run-make-fulldeps/core-no-fp-fmt-parse/Makefile) (#86048) cfg features have got their own run-make-fulldeps tests to ensure building with them still works, like a Tier-2 promise. You probably needs to add 2 run-make-fulldeps tests as well.

`#[modular_crate]` needs something RFC-ish and is out of scope here 🙂.

ojeda commented:
> The `no_rc` and `no_sync` cfg features as-is which any tests is like a Tier-3 promise. Reviewers can't possibly figure out if all these combinations are working, so those responsible for these features have to fix them up after the fact.

So my question was whether we want to promise that every single combination of `cfg` flags will work to begin with.

If yes, one option is to have flags override others automatically, but that may be surprising, especially since these are negative (e.g. someone explicitly requested to remove `rc`, but something else re-selected it).

Thus it may be reasonable to only allow a subset of combinations instead, with the constraints encoded somewhere. From what I see, the Cargo docs suggest using `compile_error!` in the source code, which is not great, but I guess we could do it.

> The [`no_global_oom_handling`](https://github.com/rust-lang/rust/tree/master/src/test/run-make-fulldeps/alloc-no-oom-handling) (#84266) and [`no_fp_fmt_parse`](https://github.com/rust-lang/rust/blob/master/src/test/run-make-fulldeps/core-no-fp-fmt-parse/Makefile) (#86048) cfg features have got their own run-make-fulldeps tests to ensure building with them still works, like a Tier-2 promise. You probably needs to add 2 run-make-fulldeps tests as well.

That is definitely a good idea, will do. However, those PRs do not have a test for every combination, unless I am missing something.

joshtriplett commented:
> So my question was whether we want to promise that every single combination of cfg flags will work to begin with.

Right now, I don't think we're prepared to promise that any of those flags work. :)

If we want to stabilize these, we'll need to make the specific stability promise clear and any dependencies explicit. Until then, I think it's fine if these only work if used as tested.

bors commented:
:umbrella: The latest upstream changes (presumably #90542) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@ojeda if you can resolve the conflicts we can get this merged

JohnCSimon commented:
@ojeda 
Ping from triage: Can you please post the status of this PR? 
Thank you.

ojeda commented:
@Dylan-DPC @JohnCSimon Thanks for the ping and apologies for the delay.

Rebased, added missing gates to allow for independent compilation of each flag, added a `run-make-fulldeps` test for each and compile-tested.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5650588110?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/4ad738053feff36d6cf081653904d90c6d1cfc71/checks/5650588110/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc-demangle v0.1.21
error: unexpected `cfg` condition name
   --> library/alloc/src/lib.rs:206:11
    |
206 | #[cfg(not(no_rc))]
    |
    |
    = note: `-D unexpected-cfgs` implied by `-D warnings`
error: unexpected `cfg` condition name
   --> library/alloc/src/lib.rs:211:15
    |
    |
211 | #[cfg(all(not(no_rc), not(no_sync), target_has_atomic = "ptr"))]

error: unexpected `cfg` condition name
   --> library/alloc/src/lib.rs:211:27
    |
    |
211 | #[cfg(all(not(no_rc), not(no_sync), target_has_atomic = "ptr"))]

error: unexpected `cfg` condition name
   --> library/alloc/src/lib.rs:213:44
    |
    |
213 | #[cfg(all(not(no_global_oom_handling), not(no_rc), not(no_sync), target_has_atomic = "ptr"))]

error: unexpected `cfg` condition name
   --> library/alloc/src/lib.rs:213:56
    |
    |
213 | #[cfg(all(not(no_global_oom_handling), not(no_rc), not(no_sync), target_has_atomic = "ptr"))]

error: could not compile `alloc` due to 5 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

bjorn3 commented:
You will need to add no_rc and no_sync to the array at https://github.com/rust-lang/rust/blob/64137f0b15b752d0c734661dc713bcd140858320/src/bootstrap/lib.rs#L197

ojeda commented:
> You will need to add no_rc and no_sync to the array at

Thanks! I took the chance to look for other instances and noticed we should put them in the `doc(cfg_hide())` too, so I did that too.

ojeda commented:
I also noticed the other `run-make-fulldeps` test in `alloc` was a bit out of date, so I sent https://github.com/rust-lang/rust/pull/95219 to align it to the ones introduced here.

Dylan-DPC commented:
r? rust-lang/libs

Dylan-DPC commented:
r? @joshtriplett

ojeda commented:
Rebased, added `-Dwarnings` to the tests and changed `-include`s to `include` according to ba74a628adfa which got merged a couple days ago.

Mark-Simulacrum commented:
@bors r+

This looks good to me; I imagine it might end up needing rebasing if we uncover a few more places we need the cfg but for now seems like a good start.

bors commented:
:pushpin: Commit 614c2e404a2eae63c2f392fd90e98960b6ec59bc has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 614c2e404a2eae63c2f392fd90e98960b6ec59bc -->
<!-- homu: {"type":"Approved","sha":"614c2e404a2eae63c2f392fd90e98960b6ec59bc","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
